src/cuda-kernels/identities.cu

```diff
@@ -0,0 +1,43 @@
+// BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/master/LICENSE
+
+#include "awkward/cuda-kernels/cuda_identities.h"
+#include <stdio.h>
+
+__global__
+void cuda_Identities32_toIdentities64(
+  int64_t* toptr,
+  const int32_t* fromptr,
+  int64_t length) {
+  int64_t block_id = blockIdx.x + blockIdx.y * gridDim.x + gridDim.x * gridDim.y * blockIdx.z;
+  int64_t thread_id = block_id * blockDim.x + threadIdx.x;
+  if(thread_id < length) {
+    toptr[thread_id] = (int64_t)(fromptr[thread_id]);
+  }
+}
+
+ERROR awkward_cuda_Identities32_to_Identities64(
+  int64_t* toptr,
+  const int32_t* fromptr,
+  int64_t length,
+  int64_t width) {
+
+  dim3 blocks_per_grid;
+  dim3 threads_per_block;
+
+  if (length > 1024) {
```

Comment:
Should this be `(length * width) > 1024` instead? (same for the other references to `length` when assigning `blocks_per_grid` and `threads_per_block`)

Reply:
Yes, you're right. It should be `length * width`. I'll put this entire thing in a utility function which could be named, `fetch_num_blocks` and `fetch_num_threads` and call that.

Reply:
It doesn't look like that's happened yet, so I won't merge right away. But for generality of utility functions, note that the three cases all do different things:

   * NumpyArray has a raw buffer of bytes; how it should be divided up into threads and blocks depends on if and how those bytes are being interpreted. Reducers interpret the bytes as numbers, but basic copy operations like `carry` only need to know byte ranges. NumpyArrays might not be contiguous: `array[:, :-1]` is an example of an operation that takes a contiguous `array` to a non-contiguous one. For some operations, dealing with non-contiguousness was too hard to think about so we run `NumpyArray::contiguous` to get it into that state before calling the kernel, but others deal with the non-contiguousness via `strides` and `shape`.
   * Index always has a compile-time integer type, is always one-dimensional, and is always contiguous. We use Indexes a lot more than NumpyArray (for example, `ak.num` doesn't need to see the underlying NumpyArray: the number of items is described by a list node's `offsets`, `starts`/`stops`, or `size`). Since we use them everywhere and don't want to overcomplicate the code, Index is forced to be contiguous.
   * Identities always has a compile-time integer type, is always two-dimensional, and is always contiguous. We also don't use them much.

Even though one of these (Identities) is two-dimensional, and some NumpyArray data are n-dimensional (if `len(shape) > 1`), I don't think we ever need to use CUDA's two or three-dimensional blocks because we never make use of n-dimensional "neighbor-ness" in any kernels. I think the two and three-dimensional blocks exist for people who are doing image analysis or PDE solving, in which memory for cell `(i, j)` needs to be close to `(i + 1, j)` and `(i, j + 1)` because they're taking differences or computing convolutions across those dimensions. Some of our (non-embarrassingly parallel) kernels need to relate cell `i` with `i + 1`, but never anything that resembles a two or three-dimensional grid.

Just in case you're thinking that we might need to do that—we don't.